examples/gno.land/r/sys/validators/v2/poc.gno

```diff
@@ -16,18 +41,54 @@ func NewPropRequest(changesFn func() []validators.Validator, title, description
 	}
 
 	callback := func(cur realm) error {
+		// Enforce cooldown: proposals can only update the validator set after the cooldown period.
+		if time.Since(lastValsetUpdate) < valsetUpdateCooldown {
+			return errors.New(errValsetUpdateCooldown)
+		}
+
+		prevTotalVotingPower := vp.TotalVotingPower()
+
+		validators := vp.GetValidatorsMap()
+
+		var addedVotingPower uint64
+
+		// Gather all proposed validator changes.
 		for _, change := range changesFn() {
+			val, exists := validators[change.Address]
+			if exists {
+				val.VotingPower = change.VotingPower
+				validators[change.Address] = val
+			} else {
+				addedVotingPower += change.VotingPower
+			}
+
 			if change.VotingPower == 0 {
-				// This change request is to remove the validator
-				removeValidator(change.Address)
+				if !exists {
+					return errors.New("cannot remove non-existing validator: " + change.Address.String())
+				}
 
+				removeValidator(change.Address)
 				continue
 			}
 
-			// This change request is to add the validator
 			addValidator(change)
 		}
 
+		var updatedVotingPower uint64
+		for _, val := range validators {
+			updatedVotingPower += val.VotingPower
+		}
+
+		newTotalVotingPower := addedVotingPower + updatedVotingPower
+		// Guardrails:
+		// 1. Total voting power cannot be zero after the update.
+		// 2. Legacy validators must retain at least the configured trust level of the voting power in the new set.
+		if prevTotalVotingPower > 0 && (newTotalVotingPower == 0 || updatedVotingPower*trustLevelRatio.denominator < newTotalVotingPower*trustLevelRatio.numerator) {
```

Comment:
Just a nitpick that this multiplication can potentially overflow, but not sure if we should handle this now, or bound it to a max voting power